[AutoDiff] Simplify conditions enabling differentiable programming.

dan-zheng opened the pull request:
Previously, two conditions were necessary to enable differentiable programming:
- Specifying the `-enable-experimental-differentiable-programming` frontend flag.
- Importing the `_Differentiation` module.

Importing the `_Differentiation` module is the true condition because it
contains the required compiler-known `Differentiable` protocol. The frontend
flag is redundant and cumbersome.

Now, the frontend flag is removed. Importing `_Differentiation` is the only condition.

dan-zheng committed:
[AutoDiff] Simplify conditions enabling differentiable programming.

Previously, two conditions were necessary to enable differentiable programming:
- Using the `-enable-experimental-differentiable-programming` frontend flag.
- Importing the `_Differentiation` module.

Importing the `_Differentiation` module is the true condition because it
contains the required compiler-known `Differentiable` protocol. The frontend
flag is redundant and cumbersome.

Now, the frontend flag is removed.
Importing `_Differentiation` is the only condition.

dan-zheng commented:
@swift-ci Please smoke test

rxwei commented on lib/Sema/TypeChecker.h:
```diff
@@ -1520,6 +1520,12 @@ bool isMemberOperator(FuncDecl *decl, Type type);
+bool isDifferentiableProgrammingEnabled(SourceFile &SF);
+
+/// Returns `true` iff `AdditiveArithmetic` derived conformances are enabled.
+bool areAdditiveArithmeticDerivedConformancesEnabled(SourceFile &SF);
```
Starting with "are" is surely more grammatically correct, but it only seems to make this API less discoverable. APIs that start with "are" are usually ones that predicate on a plural noun phrase, but the "derived conformances" functionality is almost always treated as a single entity. I'd suggest making this API start with `is`, and for that purpose, changing the subject in the phrase to a singular noun phrase, e.g. "conformance derivation".

dan-zheng replied:
Done in https://github.com/apple/swift/pull/30765/commits/51284a89b38922b1322edb34f55c516bcc05af49. I considered the same!

rxwei reviewed (approved):
Does `@_exported import _Differentiable` in a module make all clients of that module have differentiable programing enabled?  It'd be good to verify that.

dan-zheng committed:
NFC: Rename helper.

`areAdditiveArithmeticDerivedConformancesEnabled` -> `isAdditiveArithmeticConformanceDerivationEnabled`

dan-zheng commented:
> Does `@_exported import _Differentiable` in a module make all clients of that module have differentiable programing enabled? It'd be good to verify that.

Good idea! I'll work on a test for this.

dan-zheng commented:
@swift-ci Please smoke test

marcrasi commented on lib/Sema/TypeChecker.cpp:
```diff
@@ -415,6 +415,35 @@ void swift::performWholeModuleTypeChecking(SourceFile &SF) {
+bool swift::isDifferentiableProgrammingEnabled(SourceFile &SF) {
+  auto &ctx = SF.getASTContext();
+  // Return true if differentiable programming is explicitly enabled.
+  if (ctx.LangOpts.EnableExperimentalDifferentiableProgramming)
```
Does this condition ever trigger, now that the frontend flag is gone?

marcrasi commented on lib/Sema/TypeChecker.cpp:
```diff
@@ -415,6 +415,35 @@ void swift::performWholeModuleTypeChecking(SourceFile &SF) {
+      break;
+    }
+  }
+  return importsDifferentiationModule;
```
Is this `true` for the `_Differentiation` module itself? It seems like it must be, otherwise the `_Differentiation` module wouldn't compile successfully. But it's a bit surprising that `namelookup::getAllImports` returns the current module.

marcrasi reviewed (commented):
I posted a few inline questions about how this works :)
